controllers/metal3.io/hostfirmwarecomponents_controller.go

```diff
@@ -0,0 +1,319 @@
+
+// Reconcile handles changes to HostFirmwareComponents resources.
+func (r *HostFirmwareComponentsReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+	reqLogger := r.Log.WithValues("hostfirmwarecomponents", req.NamespacedName)
+	reqLogger.Info("start")
+
+	// Get the corresponding baremetalhost in this namespace, if one doesn't exist don't continue processing
+	bmh := &metal3api.BareMetalHost{}
+	err = r.Get(ctx, req.NamespacedName, bmh)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		reqLogger.Info("could not get baremetal host, not running hostfirmwarecomponents reconciler")
+		return ctrl.Result{Requeue: true, RequeueAfter: resourceNotAvailableRetryDelay}, nil
+	}
+
+	if hasDetachedAnnotation(bmh) {
+		reqLogger.Info("the host is detached, not running hostfirmwarecomponents reconciler")
+		return ctrl.Result{Requeue: true, RequeueAfter: unmanagedRetryDelay}, nil
+	}
+	// If the reconciliation is paused, requeue
+	annotations := bmh.GetAnnotations()
+
+	if _, ok := annotations[metal3api.PausedAnnotation]; ok {
+		reqLogger.Info("host is paused, no work to do")
+		return ctrl.Result{Requeue: true, RequeueAfter: subResourceNotReadyRetryDelay}, nil
+	}
+
+	// Fetch the HostFirmwareComponents
+	hfc := &metal3api.HostFirmwareComponents{}
+	info := &rhfcInfo{ctx: ctx, log: reqLogger, hfc: hfc, bmh: bmh}
+	if err = r.Get(ctx, req.NamespacedName, hfc); err != nil {
+		// The HFC resource may have been deleted
+		if k8serrors.IsNotFound(err) {
+			reqLogger.Info("hostFirmwareComponnets not found")
+			return ctrl.Result{Requeue: false}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, fmt.Errorf("could not load hostFirmwareComponents: %w", err)
+	}
+
+	// Create a provisioner to access Ironic API
+	prov, err := r.ProvisionerFactory.NewProvisioner(ctx, provisioner.BuildHostDataNoBMC(*bmh), info.publishEvent)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create provisioner: %w", err)
+	}
+
+	ready, err := prov.TryInit()
+	if err != nil || !ready {
+		var msg string
+		if err == nil {
+			msg = "not ready"
+		} else {
+			msg = err.Error()
+		}
+		reqLogger.Info("provisioner is not ready", "Error", msg, "RequeueAfter", provisionerRetryDelay)
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerRetryDelay}, nil
+	}
+
+	newStatus, err := r.updateHostFirmware(info)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("could not update hostfirmwarecomponents: %w", err)
+	}
+
+	// Check ironic for the components information if possible
+	components, err := prov.GetFirmwareComponents()
+	info.log.Info("retrieving firmware components and saving to resource", "node", bmh.Status.Provisioning.ID)
+
+	if err != nil {
+		reqLogger.Error(err, "provisioner returns error", "RequeueAfter", provisionerRetryDelay)
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerRetryDelay}, nil
```

Comment:
Fun fact: I've spent the last 10 minutes figuring out when to return an error and when just to retry. I'm still not 100% sure :) Possibly, we need to return an error here to ensure proper logging and ensure an exponential backoff?

Reply:
I'm a bit lost here,
When you say return an error itsn't this cover by:
```
reqLogger.Error(err, "provisioner returns error", "RequeueAfter", provisionerRetryDelay)
```
exponential backoff you mean the value for `RequeueAfter` ?

Reply:
When you do something like:
```go
return ctrl.Result{}, someError
```
The controller runtime will log the error and retry with an exponential backoff.

If you do
```go
return ctrl.Result{Retry: true}, nil
```
the controller runtime will also do exponential backoff, as far as I could understand. But I could not understand well :)

In any case, it feels logical that when we hit an error somewhere internally (i.e. not clearly caused by a user), we need to return this error to the controller runtime.